sdks/python/apache_beam/dataframe/frames.py

```diff
@@ -1393,9 +1430,16 @@ def fill_dataframe(*args):
 
 
 
-  cummax = cummin = cumsum = cumprod = frame_base.wont_implement_method(
-      'order-sensitive')
-  diff = frame_base.wont_implement_method('order-sensitive')
+  cummax = frame_base.order_sensitive_method(pd.DataFrame, 'cummax')
+  cummin = frame_base.order_sensitive_method(pd.DataFrame, 'cummin')
+  cumprod = frame_base.order_sensitive_method(pd.DataFrame, 'cumprod')
+  cumsum = frame_base.order_sensitive_method(pd.DataFrame, 'cumsum')
+  diff = frame_base.order_sensitive_method(pd.DataFrame, 'diff')
```

Comment:
it seems that for diff, each element only depends on the previous element, not on *all previous elements*, so it may be possible to implement efficiently. Do you think it makes sense to add a bug/TODO to implement diff?

Reply:
The problem isn't that the individual operations are difficult to make efficient, it's that the underlying PCollections have no concept of ordering. So at execution time we don't know which row is supposed to be before this one. In all likelihood it's on another worker due to our hash-based partitioning.

@robertwb and I have talked about implementing order-sensitive methods in the future (I recently filed BEAM-12129 to track this). I think the way it would work is that we would allow users to perform an operation like `sort_values` that _imposes_ an ordering, and then it would be possible to perform order-sensitive operations on the output.

Under the hood we would partition "sorted" dataframes based on their ordering, rather than randomly with the hash-based partitioning.

Reply:
hmmm I see... I was thinking that you would do `index = index - 1` on the DF, and then operate the result against the original DF, but this is not possible because indices are not necessarily integers, and may not be sorted integers either, right?

Reply:
Ahh I see. Yeah I think you're right we could do that and it wouldn't be too expensive. The problem there is it would be a divergence from pandas, which always ignores the index and computes the diff based on the order of the data. That feels like a nitpicky argument, but it could be surprising for someone very familiar with pandas, so we've opted to just disallow cases like this.

There are some cases where we've added Beam-only arguments that make operations feasible in Beam (e.g. `nlargest(keep='any')`, `unique(as_series=True)`). We could consider that here, something like `diff(by_index=True)`.
I filed BEAM-12171 to track this idea, thanks :)